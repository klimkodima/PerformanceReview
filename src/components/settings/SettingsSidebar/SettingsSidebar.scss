.settings-sidebar {
  padding: 15px;

  ul {
    margin: 0;
    padding: 0;
  }
}
