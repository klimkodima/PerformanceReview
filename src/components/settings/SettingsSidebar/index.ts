export { default as SettingsSidebar } from './SettingsSidebarContainer';
