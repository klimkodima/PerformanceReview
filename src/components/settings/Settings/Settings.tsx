import { ReactElement, memo } from 'react';

import { SETTINGS_PERMISSION } from 'src/constants';
import { SettingsSidebar } from '../SettingsSidebar';

import { SettingsPermissionType } from 'src/store/App';
import { SettingsTabsType } from '../types';

type SettingsPropsType = {
  listOfTabs: SettingsTabsType[];
  onSettingsTabClick: (name: string) => void;
  settingsPermission: SettingsPermissionType;
};

const Settings = memo(
  ({
    listOfTabs,
    onSettingsTabClick,
    settingsPermission
  }: SettingsPropsType): ReactElement => (
    <>
      <SettingsSidebar
        listOfTabs={listOfTabs}
        onSettingsTabClick={onSettingsTabClick}
      />
      {listOfTabs
        .filter(({ isActive }) => isActive)
        .map(({ name, component, editableComponent }) => (
          <div key={name} className='settings-container'>
            <h2>{name}</h2>
            {settingsPermission === SETTINGS_PERMISSION.WRITE
              ? editableComponent
              : component}
          </div>
        ))}
    </>
  )
);

export default Settings;

Settings.displayName = 'Settings';
