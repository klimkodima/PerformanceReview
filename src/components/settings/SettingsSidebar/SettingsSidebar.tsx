import { ReactElement, memo } from 'react';

import { SidebarItem } from '../SidebarItem';

import { SettingsTabsType } from '../types';

import './SettingsSidebar.scss';

type SettingsSidebarPropsType = {
  listOfTabs: SettingsTabsType[];
  onSettingsTabClick: (name: string) => void;
};

const SettingsSidebar = memo(
  ({
    listOfTabs,
    onSettingsTabClick
  }: SettingsSidebarPropsType): ReactElement => (
    <div className='settings-sidebar sidebar-wrapper'>
      <ul>
        {listOfTabs.map(({ name, isActive }) => (
          <SidebarItem
            key={name}
            name={name}
            isActive={isActive}
            onSettingsTabClick={onSettingsTabClick}
          />
        ))}
      </ul>
    </div>
  )
);

export default SettingsSidebar;

SettingsSidebar.displayName = 'SettingsSidebar';
