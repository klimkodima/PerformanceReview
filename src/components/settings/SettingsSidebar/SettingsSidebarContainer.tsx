import { FC, ReactElement } from 'react';

import SettingsSidebar from './SettingsSidebar';

import { SettingsTabsType } from '../types';

type SettingsSidebarContainerPropsType = {
  listOfTabs: SettingsTabsType[];
  onSettingsTabClick: (name: string) => void;
};

const SettingsSidebarContainer: FC<SettingsSidebarContainerPropsType> = ({
  listOfTabs,
  onSettingsTabClick
}): ReactElement => (
  <SettingsSidebar
    listOfTabs={listOfTabs}
    onSettingsTabClick={onSettingsTabClick}
  />
);

export default SettingsSidebarContainer;
