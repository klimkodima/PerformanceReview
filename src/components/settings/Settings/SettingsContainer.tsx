import { ReactElement, useCallback, useState } from 'react';

import { t } from 'i18next';

import { selectSettingsPermission } from 'src/store/App';
import Settings from './Settings';
import { TasksCoefficientsTable } from '../TasksCoefficientsTable';
import {
  HoursCorrespondence,
  HoursCorrespondenceEdit
} from '../HoursCorrespondence';
import { SeniorityBonuses, SeniorityBonusesEdit } from '../SeniorityBonuses';
import {
  PresetWorkingHours,
  PresetWorkingHoursEdit
} from '../PresetWorkingHours';
import { UsersContainer } from '../Users';

import { SettingsTabsType } from '../types';
import { useSelector } from 'react-redux';

export const initListOfTabs: SettingsTabsType[] = [
  {
    name: t('settings.tasks_title'),
    component: <TasksCoefficientsTable />,
    editableComponent: <TasksCoefficientsTable />,
    isActive: false
  },
  {
    name: t('settings.hours_title'),
    component: <HoursCorrespondence />,
    editableComponent: <HoursCorrespondenceEdit />,
    isActive: false
  },
  {
    name: t('settings.preset_working_hours_title'),
    component: <PresetWorkingHours />,
    editableComponent: <PresetWorkingHoursEdit />,
    isActive: false
  },
  {
    name: t('settings.seniority_bonuses_title'),
    component: <SeniorityBonuses />,
    editableComponent: <SeniorityBonusesEdit />,
    isActive: false
  },
  {
    name: t('settings.users_title'),
    component: <UsersContainer />,
    editableComponent: <UsersContainer />,
    isActive: false
  }
];

const SettingsContainer = (): ReactElement => {
  const settingsPermission = useSelector(selectSettingsPermission);

  const [listOfTabs, setListOfTabs] =
    useState<SettingsTabsType[]>(initListOfTabs);

  const handleSettingsTabClick = useCallback((name: string) => {
    setListOfTabs(
      listOfTabs.map((tab) =>
        tab.name === name
          ? { ...tab, isActive: true }
          : { ...tab, isActive: false }
      )
    );
  }, []);

  return (
    <Settings
      settingsPermission={settingsPermission}
      listOfTabs={listOfTabs}
      onSettingsTabClick={handleSettingsTabClick}
    />
  );
};

export default SettingsContainer;
