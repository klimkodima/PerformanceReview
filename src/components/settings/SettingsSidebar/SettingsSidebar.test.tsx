import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import SettingsSidebar from './SettingsSidebar';
import { initListOfTabs } from '../Settings/SettingsContainer';

describe('SettingsSidebar', () => {
  test('renders correctly', () => {
    const tree = renderer
      .create(
        <SettingsSidebar
          listOfTabs={initListOfTabs}
          onSettingsTabClick={() => true}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('only first item should be active', () => {
    render(
      <SettingsSidebar
        listOfTabs={initListOfTabs}
        onSettingsTabClick={() => true}
      />
    );
    expect(screen.getAllByTestId('sidebar-list-item')[0]).toHaveClass('active');
    expect(screen.getAllByTestId('sidebar-list-item')[1]).not.toHaveClass(
      'active'
    );
    expect(screen.getAllByTestId('sidebar-list-item')[2]).not.toHaveClass(
      'active'
    );
    expect(screen.getAllByTestId('sidebar-list-item')[3]).not.toHaveClass(
      'active'
    );
  });

  test('renders 4 items', () => {
    render(
      <SettingsSidebar
        listOfTabs={initListOfTabs}
        onSettingsTabClick={() => true}
      />
    );
    expect(screen.getAllByTestId('sidebar-list-item').length).toBe(5);
  });
});
