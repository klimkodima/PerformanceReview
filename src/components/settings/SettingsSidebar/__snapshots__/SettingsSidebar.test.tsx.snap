// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsSidebar renders correctly 1`] = `
<div
  className="settings-sidebar sidebar-wrapper"
>
  <ul>
    <li
      className="sidebar-list-item active"
      data-testid="sidebar-list-item"
      onClick={[Function]}
    >
      Tasks' Coefficients
    </li>
    <li
      className="sidebar-list-item "
      data-testid="sidebar-list-item"
      onClick={[Function]}
    >
      Hours' Correspondence
    </li>
    <li
      className="sidebar-list-item "
      data-testid="sidebar-list-item"
      onClick={[Function]}
    >
      Preset Working Hours
    </li>
    <li
      className="sidebar-list-item "
      data-testid="sidebar-list-item"
      onClick={[Function]}
    >
      Seniority Bonuses
    </li>
    <li
      className="sidebar-list-item "
      data-testid="sidebar-list-item"
      onClick={[Function]}
    >
      Users
    </li>
  </ul>
</div>
`;
